File: 1.2. Haskell/Test-4.hs
import System.IO

isPow2 :: Int -> Bool
isPow2 1 = True
isPow2 n = n > 0 && even n && isPow2 (n `div` 2)

splitPow2 :: [Int] -> ([Int], [Int])
splitPow2 xs = (filter isPow2 xs, filter (not . isPow2) xs)

main :: IO ()
main = do
    hSetEncoding stdout utf8
    putStrLn "Test 4 (hard): large powers of 2 among other numbers"
    let xs = [64,100,128,200,256]
    putStrLn $ "List:       " ++ show xs
    putStrLn $ "Result:     " ++ show (splitPow2 xs)
    putStrLn   "Expected:   ([64,128,256],[100,200])"

File: 1.2. Prolog/Test-4.pl
:- consult('Task-1.2.pl').
:- initialization(main, main).

main :-
    write('Test 4 (hard): large powers of 2 among other numbers'), nl,
    split_pow2([64,100,128,200,256], Pow, NotPow),
    write('Powers of 2: '), write(Pow), nl,
    write('Others:      '), write(NotPow), nl,
    write('Expected:    [64,128,256] and [100,200]'), nl.
